zio/jsonio/marshal.go

```diff
@@ -0,0 +1,129 @@
+package jsonio
+
+import (
+	"encoding/hex"
+	"time"
+
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/zcode"
+	"github.com/brimdata/zed/zson"
+)
+
+func Marshal(val *zed.Value) interface{} {
+	return marshalAny(val.Type, val.Bytes)
+}
+
+type zedErr struct {
+	Error interface{} `json:"error"`
+}
+
+func marshalAny(typ zed.Type, bytes zcode.Bytes) interface{} {
+	if bytes == nil {
+		return nil
+	}
+	switch typ := typ.(type) {
+	case *zed.TypeAlias:
+		return marshalAny(typ.Type, bytes)
+	case *zed.TypeOfUint8, *zed.TypeOfUint16, *zed.TypeOfUint32, *zed.TypeOfUint64:
+		return zed.DecodeUint(bytes)
+	case *zed.TypeOfInt8, *zed.TypeOfInt16, *zed.TypeOfInt32, *zed.TypeOfInt64:
+		return zed.DecodeInt(bytes)
+	case *zed.TypeOfDuration:
+		return zed.DecodeDuration(bytes).String()
+	case *zed.TypeOfTime:
+		return zed.DecodeTime(bytes).Time().Format(time.RFC3339Nano)
+	case *zed.TypeOfFloat32:
+		return zed.DecodeFloat32(bytes)
+	case *zed.TypeOfFloat64:
+		return zed.DecodeFloat64(bytes)
+	case *zed.TypeOfBool:
+		return zed.DecodeBool(bytes)
+	case *zed.TypeOfBytes:
+		return "0x" + hex.EncodeToString(bytes)
+	case *zed.TypeOfString:
+		return string(bytes)
+	case *zed.TypeOfIP:
+		return zed.DecodeIP(bytes).String()
+	case *zed.TypeOfNet:
+		return zed.DecodeNet(bytes).String()
+	case *zed.TypeOfNull:
+		return nil
+	case *zed.TypeRecord:
+		return marshalRecord(typ, bytes)
+	case *zed.TypeArray:
+		return marshalArray(typ, bytes)
+	case *zed.TypeSet:
+		return marshalSet(typ, bytes)
+	case *zed.TypeMap:
+		return marshalMap(typ, bytes)
+	case *zed.TypeUnion:
+		return marshalUnion(typ, bytes)
+	case *zed.TypeEnum:
+		return marshalEnum(typ, bytes)
+	case *zed.TypeError:
+		return &zedErr{marshalAny(typ.Type, bytes)}
```

Comment:
Nit: Clearer I think to drop zedErr and do this.
```suggestion
		return map[string]interface{}{"error": marshalAny(typ.Type, bytes)}
```

Reply:
It seems you should use map[string]interface{} instead of struct when the keys are dynamic/unknown but ok.